### PR TITLE
Add SECURITY.md

### PR DESCRIPTION
**Pull request by NSeydoux**
This adds a file documebnting our security guidelines, and links to the Inrupt security page for reporting vulnerabilities.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -66,6 +66,7 @@ The `solid-client-authn` libraries are compatible with [NSS](https://github.com/
 - [Using Inrupt Solid JavaScript Client Libraries to authenticate](https://docs.inrupt.com/developer-tools/javascript/client-libraries/tutorial/authenticate/)
 - [Inrupt documentation Homepage](https://docs.inrupt.com/)
 - [Architecture and design documentation](./ARCHITECTURE.md)
+- [Security policies, and vulnerabilities reporting](./SECURITY.md)
```
Comment (matthieubosquet) [inline README.md:69]:
```suggestion
- [Security policy and vulnerability reporting](./SECURITY.md)
```

File: SECURITY.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Security policy
+
+This document gathers security-related policies and guidelines for the codebase
+available in this repository.
+
+Authentication is a sensitive domain, and as such we designed the `@inrupt/solid-client-authn-*`
+libraries are designed with a particular attention to security. In particular,
```
Comment (matthieubosquet) [inline SECURITY.md:7]:
```suggestion
libraries with a particular attention to security. In particular, 
```

File: SECURITY.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# Security policy
+
+This document gathers security-related policies and guidelines for the codebase
+available in this repository.
+
+Authentication is a sensitive domain, and as such we designed the `@inrupt/solid-client-authn-*`
+libraries are designed with a particular attention to security. In particular, 
+we decided to apply the following rules:
```
Comment (matthieubosquet) [inline SECURITY.md:8]: I would maybe use a numbered list in this context, what do you think?

File: SECURITY.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+- Comply with the [OAuth security guidelines](https://tools.ietf.org/id/draft-ietf-oauth-security-topics-15.html)
```
Comment (matthieubosquet) [inline SECURITY.md:10]:
```suggestion
- Comply with the [OAuth security guidelines](https://tools.ietf.org/id/draft-ietf-oauth-security-topics-15.html).
```

File: SECURITY.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+where applicable. This involves, among other things:
```
Comment (matthieubosquet) [inline SECURITY.md:11]:
```suggestion
This involves, among other things:
```

File: SECURITY.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  - No support for the implicit grant and the resource owner password grant
+  - The use of a PKCE token
+  - Binding tokens to a DPoP key to make them sender-constrained whenever possible
+- Short-lived tokens (e.g., ID token and Access token) and private cryptographic
+material (e.g. a DPoP private key) are only stored in a function closure, so that
+they may not be extracted once received from the remote server. These protected 
+elements will not be made available directly (e.g. through a function call) or
+indirectly (e.g. allowing to sign a header with the DPoP key) to any third-party.
+- Longer-lived tokens (e.g. refresh tokens) are only stored in a secure storage,
+i.e., a storage which isn't shared beyond the librarie's control. This excludes
```
Comment (matthieubosquet) [inline SECURITY.md:21]:
```suggestion
i.e., a storage which isn't shared beyond the librarie's control. This means not using
```

File: SECURITY.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+browser storage such as `localStorage` and `sessionStorage`.
```
Comment (matthieubosquet) [inline SECURITY.md:22]:
```suggestion
the `localStorage` and `sessionStorage` in browsers.
```

File: SECURITY.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+- If these restrictions on the high-level layer are blocking some legitimate use
```
Comment (matthieubosquet) [inline SECURITY.md:23]:
```suggestion
- These restrictions on the high-level API might block some legitimate use cases.
```

File: SECURITY.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+cases, the options is added when possible for advanced users to drop down to a
```
Comment (matthieubosquet) [inline SECURITY.md:24]:
```suggestion
Advanced users can drop down to a lower-level API where token and key management is left
```

File: SECURITY.md [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+lower-level API where token and key management is left to the dependant, and no
```
Comment (matthieubosquet) [inline SECURITY.md:25]:
```suggestion
to the dependant, and no longer a concern of the library.
```

File: SECURITY.md [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+longer a concern of the library.
```
Comment (matthieubosquet) [inline SECURITY.md:26]:
```suggestion
```

File: SECURITY.md [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+browser storage such as `localStorage` and `sessionStorage`.
+- If these restrictions on the high-level layer are blocking some legitimate use
+cases, the options is added when possible for advanced users to drop down to a
+lower-level API where token and key management is left to the dependant, and no
+longer a concern of the library. 
+
+# Reporting a vulnerability
+
+If you discover a vulnerability in our code, or experience a bug related to security,
+please report it following the instructions provided on our [security page](https://inrupt.com/security/).
```
Comment (matthieubosquet) [inline SECURITY.md:31]:
```suggestion
please report it following the instructions provided on [Inrupt’s security page](https://inrupt.com/security/).
```

**Review by matthieubosquet — APPROVED** [review]
It looks good.
I made a few editorial suggestions.

File: SECURITY.md (hunk 3 above)
Comment (NSeydoux) [reply SECURITY.md:8]: Haha it initially was, but then I didn't refer back to the numbers so figured out bullet points were enough

File: SECURITY.md (hunk 9 above)
Comment (NSeydoux) [reply SECURITY.md:24]: That's what we eventually want, but it's not currently the case, which is why I added "when possible" to give us the slack to actually implement this.

File: SECURITY.md [hunk 13]
```diff
@@ -18,8 +18,8 @@ they may not be extracted once received from the remote server. These protected
 elements will not be made available directly (e.g. through a function call) or
 indirectly (e.g. allowing to sign a header with the DPoP key) to any third-party.
 - Longer-lived tokens (e.g. refresh tokens) are only stored in a secure storage,
-i.e., a storage which isn't shared beyond the librarie's control. This excludes
-browser storage such as `localStorage` and `sessionStorage`.
+i.e., a storage which isn't shared beyond the librarie's control. This means not using
```
Comment (ajacksified) [inline SECURITY.md:21]:
```suggestion
i.e., a storage which isn't shared beyond the library's control. This means not using
```

File: SECURITY.md (hunk 9 above)
Comment (matthieubosquet) [reply SECURITY.md:24]:
```suggestion
A lower-level API will be added when possible where token and key management is left to the user.
```

I would maybe try formulating like the above or remove those statements altogether (until implementation catches up).
It is unclear that it's a future plan in the current formulation and I thought straight away that I'd like to check those low level APIs wondering where/what they were ^^,

**Pull request merged**